File: mobile/src/types/navigation.ts
/**
 * Navigation Types - Type definitions for navigation features
 */

export interface Location {
  lat: number;
  lng: number;
}

export interface Route {
  summary: string;
  bounds: any;
  copyrights: string;
  legs: RouteLeg[];
  overview_coordinates: number[][];
}

export interface RouteLeg {
  distance: Distance;
  duration: Duration;
  start_location: Location;
  end_location: Location;
  start_address: string;
  end_address: string;
  steps: RouteStep[];
}

export interface RouteStep {
  distance: Distance;
  duration: Duration;
  html_instructions: string;
  instructions?: string;
  maneuver?: string;
  start_location?: Location;
  end_location?: Location;
  polyline?: { points: string };
  travel_mode: string;
}

export interface Distance {
  text: string;
  value: number;
}

export interface Duration {
  text: string;
  value: number;
}

export interface NavigationStartRequest {
  route: Route;
  current_location: Location;
  destination: Location;
  navigation_preferences?: {
    voice_personality?: string;
    verbosity?: string;
    audio_priority?: string;
    announce_street_names?: boolean;
    announce_distances?: boolean;
  };
}

export interface NavigationUpdateRequest {
  current_location: Location;
  current_step_index: number;
  distance_to_next_maneuver: number;
  time_to_next_maneuver: number;
  current_speed: number;
  is_on_highway: boolean;
  approaching_complex_intersection: boolean;
  story_playing: boolean;
  audio_priority?: string;
  last_instruction_time?: Date;
}

export interface NavigationInstruction {
  text: string;
  priority: 'critical' | 'high' | 'medium' | 'low';
  timing: 'immediate' | 'prepare' | 'reminder' | 'initial' | 'confirmation';
  maneuver_type?: string;
  street_name?: string;
  exit_number?: string;
}

export interface OrchestrationAction {
  action: 'interrupt_all' | 'pause_story' | 'duck_all' | 'wait_for_gap';
  restore_after?: boolean;
  fade_duration_ms?: number;
  duck_level_db?: number;
  max_wait_seconds?: number;
}

export interface NavigationInstructionResponse {
  has_instruction: boolean;
  instruction?: NavigationInstruction;
  audio_url?: string;
  audio_duration?: number;
  orchestration_action?: OrchestrationAction;
  next_check_seconds: number;
}

export interface NavigationStateResponse {
  active: boolean;
  route_id?: string;
  current_instruction_index: number;
  last_instruction_time?: Date;
}

export interface NavigationContext {
  currentStepIndex: number;
  distanceToNextManeuver: number;
  timeToNextManeuver: number;
  currentSpeed: number;
  isOnHighway: boolean;
  approachingComplexIntersection: boolean;
  storyPlaying: boolean;
  lastInstructionTime?: Date;
}

export interface NavigationMetrics {
  distanceToNextManeuver: number;
  timeToNextManeuver: number;
  currentSpeed: number;
  isOnHighway: boolean;
  approachingComplexIntersection: boolean;
  shouldAdvanceStep: boolean;
}

File: backend/app/schemas/navigation.py
"""
Navigation Schemas - Request/Response models for navigation endpoints
"""

from pydantic import BaseModel, Field
from typing import Dict, Any, Optional, List
from datetime import datetime

from ..models.directions import Route, Location


class NavigationStartRequest(BaseModel):
    """Request to start navigation"""
    route: Route
    current_location: Location
    destination: Location
    navigation_preferences: Optional[Dict[str, Any]] = Field(
        None, 
        description="User preferences for navigation (voice, verbosity, etc.)"
    )


class NavigationUpdateRequest(BaseModel):
    """Request to update navigation based on current position"""
    current_location: Location
    current_step_index: int = Field(..., description="Index of current route step")
    distance_to_next_maneuver: float = Field(..., description="Distance in meters to next turn")
    time_to_next_maneuver: float = Field(..., description="Estimated seconds to next turn")
    current_speed: float = Field(..., description="Current speed in km/h")
    is_on_highway: bool = Field(default=False, description="Whether currently on highway")
    approaching_complex_intersection: bool = Field(
        default=False, 
        description="Whether approaching complex maneuver"
    )
    story_playing: bool = Field(default=False, description="Whether story is currently playing")
    audio_priority: Optional[str] = Field(
        default="balanced", 
        description="Audio priority mode: safety_first, balanced, story_focused"
    )
    last_instruction_time: Optional[datetime] = None


class NavigationInstructionResponse(BaseModel):
    """Response with navigation instruction"""
    has_instruction: bool
    instruction: Optional[Dict[str, Any]] = Field(
        None,
        description="Navigation instruction details if available"
    )
    audio_url: Optional[str] = Field(None, description="URL for voice instruction audio")
    audio_duration: Optional[float] = Field(None, description="Duration of audio in seconds")
    orchestration_action: Optional[Dict[str, Any]] = Field(
        None,
        description="How to orchestrate with other audio"
    )
    next_check_seconds: int = Field(
        30,
        description="Seconds until next position check"
    )


class NavigationStateResponse(BaseModel):
    """Current navigation state"""
    active: bool = Field(..., description="Whether voice navigation is active")
    route_id: Optional[str] = Field(None, description="Active route ID")
    current_instruction_index: int = Field(0, description="Current instruction index")
    last_instruction_time: Optional[datetime] = None


class NavigationPreferences(BaseModel):
    """User preferences for navigation"""
    voice_personality: Optional[str] = Field(
        None,
        description="Preferred voice for navigation"
    )
    instruction_timing: Optional[str] = Field(
        default="normal",
        description="Timing preference: early, normal, late"
    )
    verbosity: Optional[str] = Field(
        default="normal",
        description="Instruction detail level: minimal, normal, detailed"
    )
    audio_priority: Optional[str] = Field(
        default="balanced",
        description="Audio priority: safety_first, balanced, story_focused"
    )
    announce_street_names: bool = Field(
        default=True,
        description="Whether to announce street names"
    )
    announce_distances: bool = Field(
        default=True,
        description="Whether to announce distances"
    )